msal/src/test/java/com/microsoft/identity/client/e2e/tests/mocked/DeviceCodeFlowApiTest.java

```diff
@@ -0,0 +1,329 @@
+import com.microsoft.identity.client.e2e.shadows.ShadowDeviceCodeFlowCommandSuccessful;
+import com.microsoft.identity.client.e2e.shadows.ShadowDeviceCodeFlowCommandTokenError;
+import com.microsoft.identity.client.e2e.shadows.ShadowMsalUtils;
+import com.microsoft.identity.client.e2e.tests.PublicClientApplicationAbstractTest;
+import com.microsoft.identity.client.e2e.utils.RoboTestUtils;
+import com.microsoft.identity.client.exception.MsalException;
+import com.microsoft.identity.common.exception.ClientException;
+import com.microsoft.identity.common.exception.ErrorStrings;
+import com.microsoft.identity.common.internal.authorities.Authority;
+import com.microsoft.identity.common.internal.authorities.AzureActiveDirectoryAuthority;
+import com.microsoft.identity.common.internal.providers.microsoft.microsoftsts.MicrosoftStsAuthorizationErrorResponse;
+import com.microsoft.identity.common.internal.providers.microsoft.microsoftsts.MicrosoftStsAuthorizationRequest;
+import com.microsoft.identity.common.internal.providers.microsoft.microsoftsts.MicrosoftStsAuthorizationResponse;
+import com.microsoft.identity.common.internal.providers.microsoft.microsoftsts.MicrosoftStsOAuth2Configuration;
+import com.microsoft.identity.common.internal.providers.microsoft.microsoftsts.MicrosoftStsOAuth2Strategy;
+import com.microsoft.identity.common.internal.providers.microsoft.microsoftsts.MicrosoftStsTokenRequest;
+import com.microsoft.identity.common.internal.providers.oauth2.AuthorizationResult;
+import com.microsoft.identity.common.internal.providers.oauth2.OAuth2Strategy;
+import com.microsoft.identity.common.internal.providers.oauth2.OAuth2StrategyParameters;
+import com.microsoft.identity.common.internal.providers.oauth2.TokenRequest;
+import com.microsoft.identity.common.internal.providers.oauth2.TokenResult;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.UUID;
+
+import static com.microsoft.identity.internal.testutils.TestConstants.Configurations.SINGLE_ACCOUNT_DCF_TEST_CONFIG_FILE_PATH;
+
+/**
+ * Testing class for the device code flow protocol. Currently only supporting testing for the API-side
+ * of the protocol. Will be extended to test individual aspects of the flow.
+ */
+@RunWith(RobolectricTestRunner.class)
+@Config(shadows = {ShadowMsalUtils.class})
+public class DeviceCodeFlowApiTest extends PublicClientApplicationAbstractTest {
+
+    private String mUrlBody;
+    private boolean mUserCodeReceived;
+
+    @Before
+    public void setup() {
+        super.setup();
+
+        final PublicClientApplicationConfiguration config = mApplication.getConfiguration();
+        mUrlBody = ((AzureActiveDirectoryAuthority) config.getAuthorities().get(0)).getAudience().getCloudUrl();
```

Comment:
Why not `config.getDefaultAuthority().getAuthorityUril().toString()`?

Reply:
.getAuthorityUrl() is not defined here, checked it in my code
Also tried type casting to AzureActiveDirectoryAuthority

Reply:
sorry, it was .getAuthorityURL, which returns a url with /common at the end

Reply:
It probably returns /common at the end because you probably have common in your config. Even if it does, I'm not clear on why that is a problem and why we can't use that?

Reply:
We can't use /common for single-tenant apps.

Reply:
So that's essentially a developer error right? (They need to make sure they declare the correct authority in the config). In this case, it looks @t-fadura did declare the correct authority. I see his config authority is of the form `cloud/tenantId`. So `getDefaultAuthority()` should be returning that right?

Reply:
My bad. I removed my original comment because I thought I was talking about a different thing.

What Fadi was seeing is that somehow his authority has /common appended after tenant id, which is really weird. I'm not familiar with this layer, but most likely there's a bug here.

Reply:
Yup, sycned offline with @t-fadura, there's a bug here in MSAL and has been like this forever.

Reply:
Created GitHub issue: https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/1121